oeps/oep-0006.rst

```diff
@@ -0,0 +1,224 @@
+
+New named scopes
+----------------
+
+As is the situtation today, XBlock runtimes would generally not be expected to implement every possible combination of ``BlockScope`` and ``UserScope``. Instead, XBlock runtimes are encouraged to implement these new named scopes, which would be defined in the XBlock python package:
+
+* ``Scope.context_settings`` (UserScope.NONE, BlockScope.CONTEXT_TYPE)
+
+  * Example: For an LTI XBlock, a list of the LTI provider URLs and secrets used in the course
+  * Example: For a discussions XBlock: Enable or disable discussions in this course
+
+* ``Scope.context_shared_settings`` (UserScope.NONE, BlockScope.CONTEXT_ALL)
+
+  * Example: Course-specific defaults that can apply to many different blocks, such as "max_attempts" and "show_answer", which would be consumed by the Capa Problem XBlock, Drag and Drop problem XBlock, Problem Builder XBlock family, etc. (These are currently implemented via ``InheritanceMixin``, which is built-in to edx-platform and does not allow XBlock plugins to extend the list of inherited fields. However, a downside of making them a context-scoped field is that each block would require two field definitions: a ``Scope.settings`` field for the usage-specific value, and a ``Scope.context_shared_settings`` field for the default/fallback value. )
+  * Example: Submission blackout dates - don't allow users to submit answers during the specified time period.
+
+* ``Scope.context_preferences`` (UserScope.ONE, BlockScope.CONTEXT_TYPE)
+
+  * Example: Enable or disable notifications about new discussion posts in this course
+  * Example: Video playback speed for this course
+  * Example: An XBlock in which students "accumulate building blocks over the course of the semester" (`PLAT-325`_)
+
+* ``Scope.context_shared_preferences`` (UserScope.ONE, BlockScope.CONTEXT_ALL)
+
+  * Example: Allow users to self-assign themselves into groups, which affects multiple types of problem XBlock (ORA, Problem Builder)?
+  * This seems to be the least-compelling new Scope, and may be dropped from the proposal if compelling use cases cannot be found.
+
+.. _PLAT-325: https://openedx.atlassian.net/browse/PLAT-325
+
+In the Open edX platform, it is expected that Studio could implement a new UI that displays all ``Scope.context_settings`` and ``Scope.context_shared_settings`` fields from any XBlocks enabled for a given course, allowing authors to configure those settings for that course. Likewise, the LMS could implement a "Preferences" tab for each course that allows students to edit all ``Scope.context_preferences``/``Scope.context_shared_preferences`` fields for XBlocks used in that course. (Though in the future, XBlocks should be able to opt-out of these auto-generated settings UIs on a per-field basis, and instead update those settings via their own UX in views like ``studio_view``, ``student_view``, or the separately proposed ``course_view``. This is expanded upon below in "Open Questions".)
```

Comment:
I'd change this to say that "Studio will implement a new UI" because without this these new fields won't be any use to anyone.